internal/zinc-classfile/src/main/scala/sbt/internal/inc/IndexBasedZipOps.scala

```diff
@@ -0,0 +1,304 @@
+package sbt.internal.inc
+
+import java.nio.channels.{ FileChannel, Channels, ReadableByteChannel }
+import java.io._
+import java.nio.file.{ Files, Path }
+import java.util.UUID
+import java.util.zip.{ Deflater, ZipOutputStream, ZipEntry }
+
+import sbt.io.{ IO, Using }
+
+/**
+ * Provides efficient implementation of operations on zip files * that are
+ * used for implementation of the Straight to Jar feature.
+ *
+ * The implementation is based on index (aka central directory) that is
+ * located at the end of the zip file and contains among others the name/path
+ * and offset where the actual data of stored file is located. Reading zips
+ * should always be done based on that index, which means that it is often enough
+ * to manipulate this index without rewriting the other part of the file.
+ * This class heavily relies on this fact.
+ *
+ * This class abstracts over the actual operations on index i.e. reading, manipulating
+ * and storing it making it easy to replace.
+ */
+abstract class IndexBasedZipOps extends CreateZip {
+
+  /**
+   * Reads timestamps of zip entries. On the first access to a given zip
+   * it reads the timestamps once and keeps them cached for future lookups.
+   *
+   * It only supports reading stamps from a single zip. The zip passed as
+   * an argument is only used to initialize the cache and is later ignored.
+   * This is enough as stamps are only read from the output jar.
+   */
+  class CachedStampReader {
+    private var cachedNameToTimestamp: Map[String, Long] = _
```

Comment:
This will not be threadsafe unless it is volatile, probably. Alternatively, should it just be a `lazy val`?

Regardless: I agree with Jorge that if it's possible to refactor this such that either 1) stamps are read eagerly and are immutable, 2) the stamps API has external caching ... would be preferable.